src/Initializer/ParameterDB.cpp

```diff
@@ -80,6 +84,83 @@ easi::Query seissol::initializers::ElementBarycentreGenerator::generate() const
+    double quadratureWeights[NUM_QUADPOINTS];
+    seissol::quadrature::TetrahedronQuadrature(quadraturePoints, quadratureWeights, QUAD_DEG);
+    // Initialize const class members with results
+    std::copy(std::begin(quadratureWeights), std::end(quadratureWeights), std::begin(m_quadratureWeights));
+    for (int i = 0; i < NUM_QUADPOINTS; ++i) {
+      std::copy(std::begin(quadraturePoints[i]), std::end(quadraturePoints[i]), std::begin(m_quadraturePoints[i]));
+    }
+    // Initialize element volumes
+    m_elemVolumes = elementVolumes();
+  }
+
+easi::Query seissol::initializers::ElementAverageGenerator::generate() const {
+  std::vector<Element> const& elements = m_meshReader.getElements();
+  std::vector<Vertex> const& vertices = m_meshReader.getVertices();
+
+  // Generate query using subpoints for each element
+  easi::Query query(elements.size() * NUM_QUADPOINTS, 3);
+  
+  // Transform subpoints to global coordinates for all elements
+  for (unsigned elem = 0; elem < elements.size(); ++elem) {
+    for (unsigned i = 0; i < NUM_QUADPOINTS; ++i) {
+      std::array<double, 3> xyz{};
+      seissol::transformations::tetrahedronReferenceToGlobal(vertices[ elements[elem].vertices[0] ].coords, vertices[ elements[elem].vertices[1] ].coords,
+        vertices[ elements[elem].vertices[2] ].coords, vertices[ elements[elem].vertices[3] ].coords, m_quadraturePoints[i].data(), xyz.data());
+      for (unsigned dim = 0; dim < 3; ++dim) {
+        query.x(elem * NUM_QUADPOINTS + i,dim) = xyz[dim];
+      }
+      // Group
+      query.group(elem * NUM_QUADPOINTS + i) = elements[elem].group;
+    }
+  }
+
+  return query;
+}
+
+std::vector<double> seissol::initializers::ElementAverageGenerator::elementVolumes() {
+  std::vector<Element> const& elements = m_meshReader.getElements();
+  std::vector<Vertex> const& vertices = m_meshReader.getVertices();
+  std::vector<double> elemVolumes(elements.size(), 0.0);
+  std::array<double, 3> a{};
+  std::array<double, 3> b{};
+  std::array<double, 3> c{};
+  std::array<double, 3> bxc{};
+
+  // Compute tetrahedron volumes as 1/6 * |triple product| = 1/6 * |a • (b x c)|
+  // where a := (v_1 − v_0) , b := (v_2 − v_0) and c := (v_3 − v_0)
+  for (unsigned elem = 0; elem < elements.size(); ++elem) {
+    for (int i = 0; i < 3; ++i) {
+      a[i] = vertices[ elements[elem].vertices[1] ].coords[i] - vertices[ elements[elem].vertices[0] ].coords[i];
+      b[i] = vertices[ elements[elem].vertices[2] ].coords[i] - vertices[ elements[elem].vertices[0] ].coords[i];
+      c[i] = vertices[ elements[elem].vertices[3] ].coords[i] - vertices[ elements[elem].vertices[0] ].coords[i];
+    }
+
+    // Cross product
+    bxc[0] = b[1] * c[2] - b[2] * c[1];
+    bxc[1] = b[2] * c[0] - b[0] * c[2];
+    bxc[2] = b[0] * c[1] - b[1] * c[0];
+
+    // Dot product
+    for (int i = 0; i < 3; ++i) {
+      elemVolumes[elem] += a[i] * bxc[i];
+    }
+    elemVolumes[elem] = abs(elemVolumes[elem]) / 6;
```

Comment:
This should be replaced with the functions in MeshTools:
https://github.com/SeisSol/SeisSol/blob/master/src/Geometry/MeshTools.h#L184